tests/test_updater_top_level_update.py

```diff
@@ -0,0 +1,462 @@
+
+        self.sim.timestamp.version = 1
+        with self.assertRaises(ReplayedMetadataError):
+            self._run_refresh()
+
+        md_timestamp = Metadata.from_file(
+            os.path.join(self.metadata_dir, "timestamp.json")
+        )
+        self.assertEqual(md_timestamp.signed.version, 2)
+
+    def test_new_timestamp_snapshot_rollback(self) -> None:
+        # Check for a rollback attack.
+        self.sim.snapshot.version = 2
+        self.sim.update_timestamp()  # timestamp v2
+        self._run_refresh()
+
+        # Snapshot meta version is smaller than previous
+        self.sim.timestamp.snapshot_meta.version = 1
+        self.sim.timestamp.version += 1  # timestamp v3
+
+        with self.assertRaises(ReplayedMetadataError):
+            self._run_refresh()
+
+        md_timestamp = Metadata.from_file(
+            os.path.join(self.metadata_dir, "timestamp.json")
+        )
+        self.assertEqual(md_timestamp.signed.version, 2)
+
+    def test_new_timestamp_expired(self) -> None:
+        # Check for a freeze attack
+        self.sim.timestamp.expires = self.past_datetime
+        self.sim.update_timestamp()
+
+        with self.assertRaises(ExpiredMetadataError):
+            self._run_refresh()
+
+        self._assert_files_exist(["root"])
+
+    def test_new_snapshot_hash_mismatch(self) -> None:
+        # Check against timestamp role’s snapshot hash
+
+        # Update timestamp with snapshot's hashes
+        self.sim.compute_metafile_hashes_length = True
+        self.sim.update_timestamp()  # timestamp v2
+        self._run_refresh()
+
+        # Modify snapshot contents without updating
+        # timestamp's snapshot hash
+        self.sim.snapshot.expires += timedelta(days=1)
+        self.sim.snapshot.version += 1  # snapshot v2
+        self.sim.timestamp.snapshot_meta.version = self.sim.snapshot.version
+        self.sim.timestamp.version += 1  # timestamp v3
+
+        # Hash mismatch error
+        with self.assertRaises(RepositoryError):
+            self._run_refresh()
+
+        md_timestamp = Metadata.from_file(
+            os.path.join(self.metadata_dir, "timestamp.json")
+        )
+        self.assertEqual(md_timestamp.signed.version, 3)
+
+        md_snapshot = Metadata.from_file(
+            os.path.join(self.metadata_dir, "snapshot.json")
+        )
+        self.assertEqual(md_snapshot.signed.version, 1)
+
+    def test_new_snapshot_unsigned(self) -> None:
+        # Check for an arbitrary software attack
+        self.sim.signers["snapshot"].clear()
+        with self.assertRaises(UnsignedMetadataError):
+            self._run_refresh()
+
+        self._assert_files_exist(["root", "timestamp"])
+
+    # TODO: RepositorySimulator works always with consistent snapshot
+    # enabled which forces the client to look for the snapshot version
+    # written in timestamp (which leads to "Unknown snapshot version").
+    # This fails the test for a snapshot version mismatch.
```

Comment:
Did it in e9e5965. This means that consistent_snapshot with non-root metadata means nothing for the simulator, though.

Reply:
well, it still supports consistent_snapshot... but only the latest/current consistent snapshot: trying to download using previous snapshot versions will fail in interesting ways (because the metadata will be incorrect). I think this is fine?